tests/unit_tests/CMakeLists.txt

```diff
@@ -25,18 +25,17 @@ include(${Catch2_SOURCE_DIR}/contrib/Catch.cmake)
 ### Tests                                           ###
 #######################################################
 
-if (MSVC)
-    # Moving targets into current directory
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_DEBUG ${CMAKE_CURRENT_BINARY_DIR})
-    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_CURRENT_BINARY_DIR})
-endif()
-
 add_executable(ut_melody ut_melody.cpp)
 target_link_libraries(ut_melody PUBLIC composerlib Catch2)
 catch_discover_tests(ut_melody)
 
 add_executable(ut_midi_stream ut_midi_stream.cpp)
 target_link_libraries(ut_midi_stream PUBLIC composerlib Catch2)
-file(COPY resources/twinkle.midi DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/resources")
-file(COPY resources/twinkle_harmonic.mid DESTINATION "${CMAKE_CURRENT_BINARY_DIR}/resources")
+
+if (MSVC)
```

Comment:
Talvez seja possível usar get_target_properties e set_target_properties para fazer um comando só que funcione para os dois. O target executável tem [RUNTIME_OUTPUT_DIRECTORY](https://cmake.org/cmake/help/latest/prop_tgt/RUNTIME_OUTPUT_DIRECTORY.html#prop_tgt:RUNTIME_OUTPUT_DIRECTORY) que define onde colocar executáveis e o [OUTPUT_NAME](https://cmake.org/cmake/help/latest/prop_tgt/OUTPUT_NAME.html#prop_tgt:OUTPUT_NAME) que define o nome do arquivo com o executável.

Reply:
Ainda não consegui encontrar essa solução. Teria como fazer um comando para os dois sem precisar tirar os arquivos do windows dos diretórios de Debug/Release?

Reply:
Tem um comando get_target_property que te dá o diretório sem ter que ser hard-coded. Mas se estiver complicado pode deixar, não é nada crucial.

Reply:
Tentei inicializar CMAKE_RUNTIME_OUTPUT_DIRECTORY para usar o get_target_property com o RUNTIME_OUTPUT_DIRECTORY, mas o retorno foi `<VAR>-NOTFOUND`:

```
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})
get_target_property(RESOURCES_DIR ut_midi_stream RUNTIME_OUTPUT_DIRECTORY_DEBUG)
```

Então configurei manualmente as variáveis com o set_target_properties, mas ele cria o diretório `/resources` fora do diretório de Debug/Release no MSVC:

```
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR})

set_target_properties(ut_midi_stream PROPERTIES
        RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
        RUNTIME_OUTPUT_DIRECTORY_DEBUG ${CMAKE_CURRENT_BINARY_DIR}/Debug
        RUNTIME_OUTPUT_DIRECTORY_RELEASE ${CMAKE_CURRENT_BINARY_DIR}/Release)

get_target_property(RESOURCES_DIR ut_midi_stream RUNTIME_OUTPUT_DIRECTORY_DEBUG)
MESSAGE("dir = '${RESOURCES_DIR}'")
message("dir-output = '${RUNTIME_OUTPUT_DIRECTORY}'")

configure_file("resources/twinkle_harmonic.mid" "${RESOURCES_DIR}/resources/twinkle_harmonic.mid" COPYONLY)
configure_file("resources/twinkle.midi" "${RESOURCES_DIR}/resources/twinkle.midi" COPYONLY)
```

Reply:
Parece que a solução é explorar as propriedades do target para ver qual a melhor. Essa solução parece boa.

https://stackoverflow.com/questions/7747857/in-cmake-how-do-i-work-around-the-debug-and-release-directories-visual-studio-2

De qualquer forma, para qualquer executável, a localização dos recursos é arbitrária. Quando a gente copia os recursos para a mesma pasta do executável, tende a ser um code smell, pois é o programa sempre pode ser executado de qualquer local no sistema. 

Ou seja, cópia dos recursos em si indica que o programa está contando com os recursos estarem no diretório local. O programa provavelmente está contando com [current_path](https://en.cppreference.com/w/cpp/filesystem/current_path) ou com o ofstream dada uma localização relativa. 

Por exemplo, 
- uma biblioteca não deveria depender de recursos, pois estes são passados por parâmetro. 
- Um executável, só depende de recursos se estiverem em locais específicos do sistema operacional (ex.: /usr/local/ para configurações, pasta local para aplicativos que aplicam algo na pasta local) ou um local que seja passado por parâmetro. Você pode configurar na IDE qual parâmetro vai passar para ele.
- Por fim, testes unitários podem receber o diretório por parâmetro ou direto com [target_compile_definitions](https://cmake.org/cmake/help/latest/command/target_compile_definitions.html) pois o local de onde o teste vai ser executado pode ser controlado.